### PR TITLE
feat: manage user update and patch using SCIM bulk command

### PR DESCRIPTION
**Pull request by leleueri**
fixes AM-3584

### CODE REVIEW

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-scim/src/main/java/io/gravitee/am/gateway/handler/scim/business/AbstractUserAction.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+/**
+ * @author Eric LELEU (eric.leleu at graviteesource.com)
+ * @author GraviteeSource Team
+ */
+@Slf4j
+public class AbstractUserAction {
```
Comment (lgw-gravitee) [inline gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-scim/src/main/java/io/gravitee/am/gateway/handler/scim/business/AbstractUserAction.java:46]: missing abstract keyword?

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-scim/src/main/java/io/gravitee/am/gateway/handler/scim/resources/users/UserEndpoint.java [hunk 2]
```diff
@@ -103,39 +102,27 @@ public void get(RoutingContext context) {
...
-            final String source = userSource(context);
+            final String userId = context.request().getParam("id");
 
             final JWT accessToken = context.get(ConstantKeys.TOKEN_CONTEXT_KEY);
             final String baseUrl = location(context.request());
             principal(accessToken)
                     .map(Optional::ofNullable)
-                    .switchIfEmpty(Maybe.just(Optional.empty()))
-                    .flatMapSingle(optPrincipal -> userService.update(userId, user, source, baseUrl, optPrincipal.orElse(null), context.get(CLIENT_CONTEXT_KEY)))
+                    .switchIfEmpty(Single.just(Optional.empty()))
```
Comment (lgw-gravitee) [inline gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-scim/src/main/java/io/gravitee/am/gateway/handler/scim/resources/users/UserEndpoint.java:123]: Single.Optional.empty sounds weird :p
maybe we can use Maybe?

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-scim/src/main/java/io/gravitee/am/gateway/handler/scim/service/impl/BulkServiceImpl.java [hunk 3]
```diff
@@ -104,8 +131,12 @@ public Single<BulkResponse> processBulkRequest(BulkRequest bulkRequest, Authenti
                 });
     }
 
+    private static String extractUserIdFromPath(BulkOperation validOperation) {
+        return validOperation.getPath().substring(validOperation.getPath().lastIndexOf("/") + 1);
+    }
+
     private Single<BulkOperation> checkOperation(BulkOperation operation) {
-        if (!(hasText(operation.getPath()) && operation.getPath().startsWith("/Users"))) {
+        if (!(hasText(operation.getPath()) && operation.getPath().startsWith(USERS_PATH))) {
```
Comment (lgw-gravitee) [inline gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-scim/src/main/java/io/gravitee/am/gateway/handler/scim/service/impl/BulkServiceImpl.java:139]: hasText && startsWith - here hasText is redundant I guess?

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-scim/src/main/java/io/gravitee/am/gateway/handler/scim/resources/users/UserEndpoint.java (hunk 2 above)
Comment (leleueri) [reply gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-scim/src/main/java/io/gravitee/am/gateway/handler/scim/resources/users/UserEndpoint.java:123]: if we are not converting the "empty" maybe in Optional, then we have to duplicate the content of the flatMap operator in a switchIfEmpty.

Wrapping the Optional into a Single Or a Maybe has no really effect as the update action is providing a Single at the end

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-scim/src/main/java/io/gravitee/am/gateway/handler/scim/business/AbstractUserAction.java (hunk 1 above)
Comment (leleueri) [reply gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-scim/src/main/java/io/gravitee/am/gateway/handler/scim/business/AbstractUserAction.java:46]: yes

**Review by lgw-gravitee — APPROVED** [review]

**Pull request merged**